posts/go-cli-from-scratch.md

```diff
@@ -0,0 +1,768 @@
+
+So now we're at a good point to start building out our subcommands.
+
+## Adding subcommands
+
+We have six subcommands to add but it'll go quickly since each command will follow the same steps:
+
+1. Define a friendly usage message
+2. Create the specific new command method
+3. Build out the core logic for the function
+4. Add a new case statement
+
+We already added the `version` case to the main method so let's start there.
+
+### Add Version command
+
+First, we need to define what the usage message should be, based on the requirements I put this together:
+
+NOTE: I didn't mention this before but the Version command will accept a single flag `--short`, if added the short version of the build info will be printed.
+
+```go
+// version.go
+var versionUsage = `Print the app version and build info for the current context.
+
+Usage: gupi version [options]
+
+Options:
+  --short  If true, print just the version number. Default false.
+`
+```
+
+Next we need to create the `NewVersionCommand()` which will return a pointer to a Command object.  This is where we will attach the usage message, define the flags, and specify the core logic:
+
+```go
+// version.go
+func NewVersionCommand() *Command {
+	cmd := &Command{
+		flags: flag.NewFlagSet("version", flag.ExitOnError),
+		Execute: versionFunc,
+	}
+
+	cmd.flags.BoolVar(&short, "short", false, "")
+	cmd.flags.Usage = func() {
+		fmt.Fprintln(os.Stderr, versionUsage)
+	}
+
+	return cmd
+}
+```
+
+As defined in the usage message this command should print the version number, build info and accept a `-short` flag.
+
+```go
+// version.go
+var (
+	build   = "???"
+	version = "???"
+	short   = false
+)
+
+var versionFunc = func(cmd *Command, args []string) {
+	if short {
+		fmt.Printf("brief version: v%s", version)
+	} else {
+		fmt.Printf("brief version: v%s, build: %s", version, build)
+	}
+	os.Exit(0)
+}
+```
+
+Nothing surprising about the `versionFunc`, it prints a message and exits with an all-clear exit code. What might've caught your eye is the ??? for the build and version variables. Those are meant as placeholder variables that will be overwritten during the build process. This allows us to specify version numbers in a config or dynamically include commit hashes on each build. We'll be using ldflags to do this, I'll go over this near the end. For now, if you want to read more check out [using ldflags to set version flags](https://www.digitalocean.com/community/tutorials/using-ldflags-to-set-version-information-for-go-applications).
+
+One point I glanced over is the `func(cmd *Command, args []string)` type for the `Execute` field. That means `Execute` is an anonymous function that can be passed in at any time or extracted into a separate variable as we did for `versionFunc`. Extracting is not necessary but is helpful for building readable code.
+
+The final step would be adding a case statement to our main method but since we already took care of that we're ready for this command.
```

Comment:
```suggestion
The final step would be adding a case statement to our main method. Since we already took care of that, we're ready for this command.
```